File: May-LeetCoding-Challenge/11-Flood-Fill/Flood-Fill.cpp
const int dx[4] = {-1, 0, 1, 0};
const int dy[4] = {0, -1, 0, 1};

class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        int n = image.size();
        if (n == 0) return image;
        int m = image[0].size();
        
        queue<pair<int,int>> q;
        q.push(make_pair(sr, sc));
        vector<vector<int>> vis(n, vector<int>(m, 0));
        vis[sr][sc] = 1;
        int c = image[sr][sc];
        image[sr][sc] = newColor;
        
        while(!q.empty()){
            auto cur = q.front();
            q.pop();
            for (int i = 0; i < 4; i++){
                int x = cur.first + dx[i];
                int y = cur.second + dy[i];
                if (0 <= x and x < n and 0 <= y and y < m and !vis[x][y] and image[x][y] == c){
                    q.push(make_pair(x, y));
                    vis[x][y] = 1;
                    image[x][y] = newColor;
                }
            }
        }
        return image;
    }
};

File: May-LeetCoding-Challenge/11-Flood-Fill/Flood-Fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, newColor) {
    if(newColor==image[sr][sc]){return image;}
    var color = image[sr][sc];
    dfs(image,new Array(sr,sc),color,newColor);
    return image;
};


var dfs = function(image,pos,color,newColor){
    var q = new Array(1);
    q[0] = pos;
    var m = pos[0];
    var n = pos[1];
    while(q.length>0){
        let p = q.shift();
        let a = p[0];
        let b = p[1];
        image[a][b]=newColor;
        if(a+1<image.length){
            if(image[a+1][b]==color){
                image[a+1][b]=newColor;
                q.push(new Array(a+1,b));
            }            
        }
        if(a-1>=0){
            if(image[a-1][b]==color){
                image[a-1][b]=newColor;
                q.push(new Array(a-1,b));
            }            
        }
        if(b+1<image[0].length){
            if(image[a][b+1]==color){
                image[a][b+1]=newColor;
                q.push(new Array(a,b+1));
            }            
        }
        if(b-1>=0){
            if(image[a][b-1]==color){
                image[a][b-1]=newColor;
                q.push(new Array(a,b-1));
            }            
        }
    }
};

